File: built-in plugins dsp/mb-pad-ambient.cpp
/**
 * MB Ambient Pad
 * Category : instrument
 * Type     : pad
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Ethereal ambient texture
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PAD_AMBIENT_H
#define MB_PAD_AMBIENT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPadAmbient : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-pad-ambient";
    static constexpr const char* PLUGIN_NAME    = "MB Ambient Pad";
    static constexpr const char* PLUGIN_TYPE    = "pad";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float space = 0.8f;  // range [0, 1]
    float volume = 0.6f;  // range [0, 1]
    };

    MbPadAmbient() = default;
    ~MbPadAmbient() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.space = std::clamp(params.space, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Ambient Pad
        return input;
    }
};

#endif // MB_PAD_AMBIENT_H

File: built-in plugins dsp/mb-pad-ambient.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbPadAmbientPlugin: PluginDefinition = {
    id: 'mb-pad-ambient', slug: 'mb-pad-ambient', name: 'MB Ambient Pad', category: 'instrument', type: 'pad', version: '1.0.0',
    description: 'Ethereal ambient texture', author: 'Max Booster',
    oscillators: [{ type: 'sine', detune: 0, gain: 0.3 }, { type: 'triangle', detune: 1200, gain: 0.25 }, { type: 'sine', detune: 1900, gain: 0.25 }, { type: 'triangle', detune: -700, gain: 0.2 }],
    envelope: { attack: 2.0, decay: 1.0, sustain: 0.9, release: 4.0 },
    parameters: [
      { id: 'space', name: 'Space', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { space: 0.8, volume: 0.6 },
  };

export default MbPadAmbientPlugin;
